### PR TITLE
ENH: add identity kwarg to frompyfunc

### LINKED ISSUE #8245: ENH: frompyfunc set user specified identity attribute
frompyfunc does not allow the ufunc's identity to be specified, and that attribute of the returned object is not writable, so it is always None.  Please add a kwarg to set the function's identity.  This is especially valuable if the ufunc will be used with the reduce family of methods.  Thank you.

It looks straightforward [here](https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/umathmodule.c#L86).  If this sounds reasonable I'll try and do it

### PR DESCRIPTION
**Pull request by mattharrigan**
PR to Close #8245.  Adding the kwarg seems pretty straightforward.  Understanding what identity actually does is much harder.  Several identity constants are defined [here](https://github.com/numpy/numpy/blob/master/numpy/core/include/numpy/ufuncobject.h#L268).  The fact that MinusOne is defined to 2 and None is defined to -1 makes my head hurt.  The ufunc getter for identity [here](https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/ufunc_object.c#L5669) seems to have fewer options than what's defined in the earlier link.  Finally when I look at the reduce methods they don't make sense to me either.  Is there any documentation defining all the possible values for identity and what they mean?

~WIP: ~ (edit by @eric-wieser: rebased and finished)

- [x] still need to edit release notes and frompyfunc doc
- [ ] May need an email to numpy discussion for the questions above or simply because of the API change.
- [x] finally should add tests for the new functionality, but may have to wait till the getter and identity values are straightened out

### CODE REVIEW

**Comment by mattharrigan** [comment]
This script helps show some of the differences between identity values.

```
import numpy as np
def f(x, y):
    print(x, y)
    return x + y

print()
print('get identity with different input')
print('None returns {}'.format(np.frompyfunc(f,2,1).identity))
for i in range(-2, 3):
    print('{} returns {}'.format(i, np.frompyfunc(f,2,1,i).identity))

x = [[1, 2],[3,4]]
print()
for i in range(-2, 3):    
    print('reduce with identity = {}'.format(i))
    print(np.frompyfunc(f,2,1,i).reduce(x))
```

**Comment by charris** [comment]
Haven't looked at this yet, but the enhancement would be useful.

**Comment by eric-wieser** [comment]
The relevant getter for `ufunc.identity` in python is [in `ufunc_get_identity`](https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/ufunc_object.c#L5668-L5680)

**Comment by eric-wieser** [comment]
>  The ufunc getter for identity here seems to have fewer options than what's defined in the earlier link

The only two that are ommited are `PyUFunc_None` and `PyUFunc_ReorderableNone`, both of which are returned as `None` in `ufunc.identity`. Perhaps there should be two special singleton values here instead, to allow python code to distingish them?

**Comment by mattharrigan** [comment]
That might make sense.  Should I also add a private setter function to convert from input to the internal constants?  I think ultimately for some input identity value inp, np.frompyfunc(f, nin, nout, inp).identity == inp (or raise an exception) for all possible values of inp.

**Comment by eric-wieser** [comment]
At the very least, it'd be nice to get the change support kwargs to `frompyfunc` in

**Comment by eric-wieser** [comment]
Seem to me that we first want to overhaul `ufunc.identity` to allow any `PyObject *`, and only then is this worth pursuing. Under the hood, we just convert them to `PyObject *`s each time we need them anyway.

I've a patch in progress in my tree, but it's unlikely to get many brain cycles in the next month. Am I right in thinking that appending fields to `Py_UfuncObject` breaks neither the API nor the ABI?`

**Comment by mattip** [comment]
#8955 has been merged, `PyUFuncObject` now has a field that can hold arbitrary `PyObject` s

@mattharrigan would you like to revive this?

**Comment by eric-wieser** [comment]
Rebased and solved conflicts. Tweaked the `identity` argument to be keyword-only.

File: numpy/core/src/umath/umathmodule.c [hunk 1]
```diff
@@ -146,10 +148,10 @@ ufunc_frompyfunc(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *NPY_UNUS
     /* Do a better job someday */
     doc = "dynamic ufunc based on a python function";
 
-    self = (PyUFuncObject *)PyUFunc_FromFuncAndData(
+    self = (PyUFuncObject *)PyUFunc_FromFuncAndDataAndSignatureAndIdentity(
             (PyUFuncGenericFunction *)pyfunc_functions, data,
-            types, /* ntypes */ 1, nin, nout, PyUFunc_None,
-            str, doc, /* unused */ 0);
+            types, /* ntypes */ 1, nin, nout, identity ? PyUFunc_IdentityValue : PyUFunc_None,
+            str, doc, /* unused */ 0, NULL, identity);
```
Comment (eric-wieser) [inline numpy/core/src/umath/umathmodule.c]: A future PR could add `signature`, which would help `vectorize`

**Comment by eric-wieser** [comment]
Test added

**Comment by seberg** [comment]
Thank Eric and @mattharrigan. The signature addition has been on my/our "Not Done" list for a bit. Should fairly straight forward, but requires a new slightly different inner loop to handle.

**Comment by seberg** [comment]
@eric-wieser feel free to merge, I think this is good. There are tiny changes in that arguments are now not only positional (which is good). I guess a ping to the mailing list can be good, but since this is a core ufunc feature exposing it to python seems very straight forward. I *feel* the argument description is a bit difficult, but I cannot figure out how to make it easier. Maybe the link to the ufunc documentation is simply enough.

**Comment by seberg** [comment]
Hmm, seems I could have just merged it back then. I think it is an obvious enough addition to have the `identity` kwarg in `frompyfunc` since it is something we have in the C-API.

**Pull request merged**